src/pocketmine/network/upnp/UPnP.php

```diff
@@ -19,63 +19,232 @@
 
 /**
- * UPnP port forwarding support. Only for Windows
+ * UPnP port forwarding support.
  */
 namespace pocketmine\network\upnp;
 
 use pocketmine\utils\Internet;
-use pocketmine\utils\Utils;
-use function class_exists;
-use function is_object;
+use SimpleXMLElement;
+use function socket_create;
+use function socket_set_option;
+use function socket_recvfrom;
+use function socket_sendto;
+use function trim;
+use function strlen;
+use function preg_match;
+use function parse_url;
+use function sprintf;
+use const AF_INET;
+use const SOCK_DGRAM;
+use const SOL_UDP;
+use const IPPROTO_IP;
+use const IP_MULTICAST_IF;
+use const IP_MULTICAST_LOOP;
+use const IP_MULTICAST_TTL;
+use const SOL_SOCKET;
+use const SO_RCVTIMEO;
 
 abstract class UPnP{
+	/** @var string|null */
+	private static $serviceURL = null;
+	
+	public static function getServiceUrl() : string{
+		$socket = @socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+		if($socket === false){
+			throw new \RuntimeException("Socket error: " . trim(socket_strerror(socket_last_error())));
+		}
+		if(!@socket_set_option($socket, SOL_SOCKET, SO_RCVTIMEO, ["sec" => 3, "usec" => 0])){
+			throw new \RuntimeException("Socket error: " . trim(socket_strerror(socket_last_error($socket))));
+		}
+		$contents =
+			"M-SEARCH * HTTP/1.1\r\n" .
+			"MX: 2\r\n" .
+			"HOST: 239.255.255.250:1900\r\n" .
+			"MAN: \"ssdp:discover\"\r\n" .
+			"ST: upnp:rootdevice\r\n\r\n";
+		$sendbyte = @socket_sendto($socket, $contents, strlen($contents), 0, "239.255.255.250", 1900);
+		if($sendbyte === false){
+			throw new \RuntimeException("Socket error: " . trim(socket_strerror(socket_last_error($socket))));
+		}
+		if($sendbyte !== strlen($contents)){
+			throw new \RuntimeException("Socket error: Unable to send the entire contents.");
+		}
+		if(@socket_recvfrom($socket, $buffer, 1024, 0, $responseHost, $responsePort) === false){
+			throw new \RuntimeException("Socket error: " . trim(socket_strerror(socket_last_error($socket))));
+		}
+		socket_close($socket);
+		$pregResult = preg_match('/location\s*:\s*(.+)\n/i', $buffer, $matches);
+		if($pregResult === false){
+			// you can use preg_last_error_msg() since PHP 8.
+			$errormsg = "";
+			switch(preg_last_error()){
+				case PREG_NO_ERROR:
+					$errormsg = "no error but preg_match returned false";
+					break;
+				case PREG_INTERNAL_ERROR:
+					$errormsg = "internal error";
+					break;
+				case PREG_BACKTRACK_LIMIT_ERROR:
+					$errormsg = "backtrack limit error";
+					break;
+				case PREG_RECURSION_LIMIT_ERROR:
+					$errormsg = "recursion limit error";
+					break;
+				case PREG_BAD_UTF8_ERROR:
+					$errormsg = "bad utf8 error";
+				break;
+				case PREG_BAD_UTF8_OFFSET_ERROR:
+					$errormsg = "bad utf8 offset error";
+					break;
+				case PREG_JIT_STACKLIMIT_ERROR:
+					$errormsg = "jit stacklimit error";
+					break;
+				default:
+					$errormsg = "unknown error";
+			}
+			throw new \RuntimeException("Preg error: " . $errormsg);
```

Comment:
This would take much less space if you do something like this: https://github.com/pmmp/PocketMine-MP/blob/15401d740f11f3a543abaecc7a449bcba3c2fc91/src/pocketmine/utils/TextFormat.php#L77-L88